### PR TITLE
Make `TransactionPool.opApply` templated

### PR DESCRIPTION
**Pull request by kchulj**
This allows us to have proper attributes being inferred on the users, as currently the delegate has no attributes.

Split off from #1161
Relates issue #1083 
Developed with @Geod24

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
Was this split off from another PR? I thought I already saw this code.

**Comment by Geod24** [comment]
You can use `Co-Authored-By: $AUTHOR` in the commit message.
But really, I was not the author, just the committer ;)
Also when splitting off PR, please add a reference, in this case #1161

File: source/agora/common/TransactionPool.d [hunk 1]
```diff
@@ -197,20 +197,38 @@ public class TransactionPool
 
     ***************************************************************************/
 
-    public int opApply (scope int delegate(Hash hash, Transaction tx) dg) @trusted
+    public int opApply (DG1T) (scope DG1T dg)
```
Comment (AndrejMitrovic) [inline source/agora/common/TransactionPool.d]: Why the name `DG1T`?

Comment (Geod24) [reply source/agora/common/TransactionPool.d]: Originally because it was the version with a single parameter (as opposed to `DG2T`).

File: source/agora/node/Ledger.d [hunk 2]
```diff
@@ -347,7 +347,7 @@ public class Ledger
 
         this.enroll_man.getEnrollments(data.enrolls,
             Height(this.getBlockHeight()));
-        foreach (hash, tx; this.pool)
+        foreach (ref Transaction tx; this.pool)
```
Comment (linked0) [inline source/agora/node/Ledger.d:350]:
```suggestion
        foreach (const ref Transaction tx; this.pool)
```

File: source/agora/node/Ledger.d [hunk 3]
```diff
@@ -379,7 +379,7 @@ public class Ledger
         const expect_height = Height(this.getBlockHeight() + 1);
         auto utxo_finder = this.utxo_set.getUTXOFinder();
 
-        foreach (tx; data.tx_set)
+        foreach (ref tx; data.tx_set)
```
Comment (linked0) [inline source/agora/node/Ledger.d:382]:
```suggestion
        foreach (const ref tx; data.tx_set)
```

File: source/agora/node/Ledger.d [hunk 4]
```diff
@@ -390,7 +390,7 @@ public class Ledger
         if (data.enrolls.length + active_enrollments < Enrollment.MinValidatorCount)
             return "Enrollment: Insufficient number of active validators";
 
-        foreach (enroll; data.enrolls)
+        foreach (ref enroll; data.enrolls)
```
Comment (linked0) [inline source/agora/node/Ledger.d]:
```suggestion
        foreach (const ref enroll; data.enrolls)
```

File: source/agora/common/TransactionPool.d (hunk 1 above)
Comment (kchulj) [reply source/agora/common/TransactionPool.d]: I think it can be just `DGT` now.

**Comment by kchulj** [comment]
@AndrejMitrovic yes, this was split off from #1161 for organization.
I will make sure to mention this in the comment.

File: source/agora/common/TransactionPool.d (hunk 1 above)
Comment (Geod24) [reply source/agora/common/TransactionPool.d]: Yes

File: source/agora/common/TransactionPool.d [hunk 5]
```diff
@@ -162,7 +162,7 @@ public class TransactionPool
     public void remove (const ref Transaction tx) @trusted
     {
         // delete inputs of transaction from the set of Input hashes
-        foreach (ref input; tx.inputs)
+        foreach (const ref input; tx.inputs)
```
Comment (Geod24) [inline source/agora/common/TransactionPool.d:165]: So technically this should be in the second commit right ?

Comment (kchulj) [reply source/agora/common/TransactionPool.d:165]: Correct, I'll rebase it :)

**Comment by Geod24** [comment]
Windows is already a PITA.

**Comment by AndrejMitrovic** [comment]
> Windows is already a PITA.

I think the issue is the timeout. The Windows build doesn't use a `parallel` loop, this slows down the tests by at least 3 minutes (on my Windows machine).

So we will need to increase the timeout.

**Review by Geod24 — APPROVED** [review]

File: source/agora/node/Ledger.d [hunk 6]
```diff
@@ -379,7 +379,7 @@ public class Ledger
         const expect_height = Height(this.getBlockHeight() + 1);
         auto utxo_finder = this.utxo_set.getUTXOFinder();
 
-        foreach (tx; data.tx_set)
+        foreach (const ref tx; data.tx_set)
```
Comment (TrustHenry) [inline source/agora/node/Ledger.d:382]: It doesn't seem relevant to the first commit.
Move to the second commit right ?

File: source/agora/node/Ledger.d [hunk 7]
```diff
@@ -390,7 +390,7 @@ public class Ledger
         if (data.enrolls.length + active_enrollments < Enrollment.MinValidatorCount)
             return "Enrollment: Insufficient number of active validators";
 
-        foreach (enroll; data.enrolls)
+        foreach (const ref enroll; data.enrolls)
```
Comment (TrustHenry) [inline source/agora/node/Ledger.d:393]: Move to the second commit right ?

Comment (kchulj) [reply source/agora/node/Ledger.d:393]: The 2nd commit was for `inputs` only and I was thinking of keeping everything else in the 1st commit.
Since this is `enroll` and not related to `tx`, it might be a good idea to move it.
What do you think?

Comment (TrustHenry) [reply source/agora/node/Ledger.d:393]: How about applying the title and contents of the second commit comprehensively in the revision?
